distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Replacer.enso

```diff
@@ -0,0 +1,144 @@
+        string_builder.toString
+
+## PRIVATE
+
+   Get the size of the Replacer LRU cache. For testing.
+get_lru_size : Integer
+get_lru_size = Replacer_Cache.getLruSize
+
+## PRIVATE
+
+   Look up a replacement string in the Replacer LRU cache. For testing.
+replacer_cache_lookup : Text -> Replacer | Nothing
+replacer_cache_lookup replacement_string = Replacer_Cache.get replacement_string
+
+## PRIVATE
+group_reference_regex = "\$(([0-9]+)|(\$)|(&)|(<([^>]+)>))"
+
+## PRIVATE
+
+   Build a replacement vector.
+
+   Parse the replacement string into an alternating series of literal
+   strings and group reference numbers.
+
+   Uses Replacement_Cache to avoid rebuilding the vector for recently used
+   replacement strings.
+build_replacement_vector_cached : Text -> Pattern_2 -> Vector Replacement ! No_Such_Group ! Too_Many_Groups
+build_replacement_vector_cached replacement_string pattern =
+    Replacer_Cache.get_or_set replacement_string _->
+        build_replacement_vector replacement_string pattern
+
+## PRIVATE
+
+   Build a replacement vector.
+
+   Parse the replacement string into an alternating series of literal
+   strings and group reference numbers.
+build_replacement_vector : Text -> Pattern_2 -> Vector Replacement ! No_Such_Group ! Too_Many_Groups
+build_replacement_vector replacement_string pattern =
+    replacement_pattern = Regex_2.compile group_reference_regex
+    it = replacement_pattern.iterator replacement_string
+
+    builder = Vector.new_builder
+    go it = case it.next of
+        Match_Iterator_Value.Next filler match next_it ->
+            replacement = parse_group_number pattern match
+            replacement.if_not_error <|
+                builder.append (Replacement.Literal filler.text)
+                builder.append replacement
+                @Tail_Call go next_it
+        Match_Iterator_Value.Last filler ->
+            builder.append (Replacement.Literal filler.text)
+    result = go it
+    if result.is_error then Error.throw result else
+        builder.to_vector
+
+## PRIVATE
+
+   Parse a capture group reference.
+
+   Arguments:
+   - pattern: the Pattern_2 used to initiate the replacement. This is used
+     to identify and validate capture groups.
+   - match: the match of the replacement string against group_reference_regex
+
+   Returns a Replacement: a group number, or, in the case of `$$`, a literal.
+
+   See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions
+parse_group_number : Pattern_2 -> Match_2 -> Replacement ! No_Such_Group ! Too_Many_Groups
+parse_group_number pattern match = case match.text.take 2 of
+    "$$" -> Replacement.Literal "$"
+    "$<" ->
+        group_name = match.text 6
+        Replacement.Substitution (pattern.lookup_group group_name)
+    "$&" -> Replacement.Substitution 0
+    _ -> 
+        n = Integer.parse <| match.text 2
+        if n < 100 then Replacement.Substitution (pattern.lookup_group n) else
```

Comment:
Where is the number `100` coming from?

Reply:
The maximum number of capture groups, from the ECMAscript spec. Changed it to a constant symbol.